File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import defaultdict, deque


def check(word1, word2):
    if len(word1) != len(word2):
        return False
    count = 0
    for w1, w2 in zip(word1, word2):
        if w1 != w2:
            count += 1
    return count == 1


def solution(begin, target, words):
    load = defaultdict(list)
    for w1 in words:
        for w2 in words:
            if w1 == w2:
                continue
            if check(w1, w2):
                load[w1].append(w2)
                load[w2].append(w1)
    for w in words:
        if check(begin, w):
            load[begin].append(w)
    
    visited = set()
    q = deque()
    q.append((begin, 0))
    visited.add(begin)
    while q:
        word, count = q.popleft()
        if word == target:
            return count
        for next_word in load[word]:
            if next_word not in visited:
                q.append((next_word, count + 1))
                visited.add(next_word)
    return 0

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    static class Word {
        String word;
        int count;
        Word(String word, int count) {
            this.word = word;
            this.count = count;
        }
    }
    static Set<String> wordSet;
    public int solution(String begin, String target, String[] words) {
        wordSet = new HashSet<>();
        int answer = bfs(begin, target, words);
        return answer;
    }
    
    private int bfs(
        String begin,
        String target,
        String[] words
    ) {
        Queue<Word> q = new LinkedList<>();
        q.offer(new Word(begin, 0));
        wordSet.add(begin);
        while (!q.isEmpty()) {
            Word word = q.poll();
            String w = word.word;
            int count = word.count;
            if (w.equals(target)) return count;
            for (String nextWord : words) {
                if (check(w, nextWord) && !wordSet.contains(nextWord)) {
                    q.offer(new Word(nextWord, count+1));
                    wordSet.add(w);
                }
            }
        }
        return 0;
    }
    
    private boolean check(String w1, String w2) {
        int diff = 0;
        for (int i = 0; i < w1.length(); i++) {
            if (w1.charAt(i) != w2.charAt(i)) diff++;
            if (diff > 1) return false;
        }
        if (diff == 1) return true;
        else return false;
    }
}